<a href="https://www.kaggle.com/code/lorresprz/santa23-permutation-identities-of-cube-puzzles?scriptVersionId=160111873" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Identities satisfied by $f,d, r$ moves of cubes $N\times N$

In this notebook, I highlight the identities satisfied by the moves of the cube puzzles. The list of identities in this notebook goes beyond those already well implemented in other notebook (such as `m*m*m = -m`). In particular, the identities presented in this notebook were found by brute force checking and generalizing the following hint

 - `f0 * f1 * r0 * (f1 ** -1) * (f0 ** -1) == d0`
 
found in this notebook 

https://www.kaggle.com/code/ryanholbrook/getting-started-with-santa-2023

for the case of $N=2$. 

The list of the first set of identities are 

- `f0 * ... * f(N-1) * rj * (f(N-1) ** -1)* ... * (f0 ** -1) == dj`
- `f0 * ... * f(N-1) * dj * (f(N-1) ** -1)* ... * (f0 ** -1) == r(N-1-j)**(-1)`
- `r0 * ... * r(N-1) * dj * (r(N-1) ** -1)* ... * (r0 ** -1) == fj`
- `r0 * ... * r(N-1) * fj * (r(N-1) ** -1)* ... * (r0 ** -1) == d(N-1-j)**(-1)`
- `d0 * ... * d(N-1) * fj * (d(N-1) ** -1)* ... * (d0 ** -1) == rj`
- `d0 * ... * d(N-1) * rj * (d(N-1) ** -1)* ... * (d0 ** -1) == f(N-1-j)**(-1)`

The list of the second set of identities can be found from the first set above:

- `(f(N-1)**-1)*...*(f0**-1) * (dj**-1) * f0 *... *f(N-1) == rj**(-1)`
- `(f(N-1)**-1)*...*(f0**-1) * r(N-1-j) * f0 *... *f(N-1) == dj**(-1)`
- `(r(N-1)**-1)*...*(r0**-1) * (fj**-1) * r0 *... *r(N-1) == dj**(-1)`
- `(r(N-1)**-1)*...*(r0**-1) * d(N-j-1) * r0 *... *r(N-1) == fj**(-1)`
- `(d(N-1)**-1)*...*(d0**-1) * (rj**-1) * d0 *... *d(N-1) == fj**(-1)`
- `(d(N-1)**-1)*...*(d0**-1) * f(N-1-j) * d0 *... *d(N-1) == rj**(-1)`

In all the formulae above

- `j = 0, ..., N-1`

Each of the above identities can be verified by running the cells below, first by inputting the value of $N$ of interest. For demonstration purpose, we choose $N=33$ in this notebook. 

In [1]:
import numpy as np
import pandas as pd
from ast import literal_eval
from pathlib import Path
from pprint import pprint
from sympy.combinatorics import Permutation

In [2]:
path = '/kaggle/input/santa-2023/'
puzzle_info = pd.read_csv(path+'puzzle_info.csv', index_col='puzzle_type')
# Parse allowed_moves
puzzle_info['allowed_moves'] = puzzle_info['allowed_moves'].apply(literal_eval)

In [3]:
# Convert allowed_moves to dict and add inverse moves
all_moves = puzzle_info.loc[:, 'allowed_moves'].to_dict()
for ptype, moves in all_moves.copy().items():
    for m, arr in moves.copy().items():
        all_moves[ptype][f"-{m}"] = np.argsort(arr).tolist()

In [4]:
#Function to extract the 6 types of cube moves: f,d,r and their inverses
def extract_allowed_moves(N):
    #N is 2-10, 19, 33
    #cube_name is : 2/2/2 to 10/10/10
    #indexlist is the range of index to extract 
    cube_name = f'cube_{N}/{N}/{N}'
    cube_moves = all_moves[cube_name]
    cl = list(cube_moves.keys())

    fp = []
    rp = []
    dp = []
    fpi = []
    rpi = []
    dpi = []
    for i in range(N):
        fp.append(Permutation(cube_moves[cl[i]]))
    for i in range(N,2*N):
        rp.append(Permutation(cube_moves[cl[i]]))
    for i in range(2*N,3*N):
        dp.append(Permutation(cube_moves[cl[i]]))
    for i in range(3*N,4*N):
        fpi.append(Permutation(cube_moves[cl[i]]))
    for i in range(4*N,5*N):
        rpi.append(Permutation(cube_moves[cl[i]]))
    for i in range(5*N,6*N):
        dpi.append(Permutation(cube_moves[cl[i]]))
    return fp,rp,dp, fpi, rpi, dpi

In [5]:
#Define commutator of 2 operators
def cm(g1, g2):
    return g1*g2*(g1**(-1))*(g2**(-1))

In [6]:
#Define a function to take the product of a list of moves
def prod(move_list):
    prod0 = 1
    for i in range(len(move_list)):
        prod0 *=move_list[i]
    return prod0

# N = 33

In [7]:
#Choose a value for N of choice, 
#here I choose N = 33 (the most complicated case for cube)
NN = 33

In [8]:
#Extract the 6 types of moves 
fp,rp,dp, fpi, rpi, dpi =  extract_allowed_moves(NN)

In [9]:
print(f'\n Identity element is Permutation({NN*NN*6-1})')


 Identity element is Permutation(6533)


 We know that for cube puzzles, $f_i$ commutes with $f_j$, $r_i$ commutes with $r_j$, $d_i$ commutes with $d_j$ for $i,j = 0, \ldots, N-1$ (all $i,j$ values). If 2 operators commute, their commutator is the identity element, which is explicitly stated above. In the case of $N=33$, this is Permutation(6533). The cell below just confirms this fact. 

In [10]:
print('\n CHECKING COMMUTATORS OF COMMUTING PAIRS')
print('##########################################################')
print(f'[fp[i]], fp[j] = Permutation({NN*NN*6-1})')
print([((i,j), cm(fp[i], fp[j])) for i in range(NN) for j in range(NN) if i<j])

#Uncomment to evaluate (very long output)
#print('##########################################################')
#print(f'[rp[i]], rp[j] = Permutation({NN*NN*6-1})')
#print([((i,j), cm(rp[i], rp[j])) for i in range(NN) for j in range(NN) if i<j])
#print('##########################################################')
#print(f'[dp[i]], dp[j] = Permutation({NN*NN*6-1})')
#print([((i,j), cm(dp[i], dp[j])) for i in range(NN) for j in range(NN) if i<j])


 CHECKING COMMUTATORS OF COMMUTING PAIRS
##########################################################
[fp[i]], fp[j] = Permutation(6533)
[((0, 1), Permutation(6533)), ((0, 2), Permutation(6533)), ((0, 3), Permutation(6533)), ((0, 4), Permutation(6533)), ((0, 5), Permutation(6533)), ((0, 6), Permutation(6533)), ((0, 7), Permutation(6533)), ((0, 8), Permutation(6533)), ((0, 9), Permutation(6533)), ((0, 10), Permutation(6533)), ((0, 11), Permutation(6533)), ((0, 12), Permutation(6533)), ((0, 13), Permutation(6533)), ((0, 14), Permutation(6533)), ((0, 15), Permutation(6533)), ((0, 16), Permutation(6533)), ((0, 17), Permutation(6533)), ((0, 18), Permutation(6533)), ((0, 19), Permutation(6533)), ((0, 20), Permutation(6533)), ((0, 21), Permutation(6533)), ((0, 22), Permutation(6533)), ((0, 23), Permutation(6533)), ((0, 24), Permutation(6533)), ((0, 25), Permutation(6533)), ((0, 26), Permutation(6533)), ((0, 27), Permutation(6533)), ((0, 28), Permutation(6533)), ((0, 29), Permutation(6533)), ((

## N=33: Identities satisfied by $f, r, d$ and their inverses

In [11]:
#############################################################
print('\n CHECKING IDENTITIES')
print(f'Test f[i=0, .. {NN}]. r[ j ]. f[i={NN},..,0]^(-1) = d[ j ] for j = 0,..,{NN}')
for k in range(NN):
	print(k, prod([fp[i] for i in range(NN)]+ [rp[k]] +[fpi[i] for i in range(NN-1,-1,-1)]) == dp[k])
#############################################################
print(f'Test f[i=0,..{NN}]. d[ j ]. f[i={NN},..,0]^(-1) = r[ {NN}-j ]^(-1) for j=0,..,{NN}')
for k in range(NN):
	print(k, prod([fp[i] for i in range(NN)]+ [dp[k]] +[fpi[i] for i in range(NN-1,-1,-1)]) == rpi[NN-1-k])
#############################################################
print(f'Test r[i=0,..{NN}]. d[ j ]. r[i={NN},..,0]^(-1) = f[ j ] for j = 0,..,{NN}')
for k in range(NN):
	print(k, prod([rp[i] for i in range(NN)]+ [dp[k]] +[rpi[i] for i in range(NN-1,-1,-1)]) == fp[k])
#############################################################
print(f'Test r[i=0,..{NN}]. f[ j ]. r[i={NN},..,0]^(-1) = d[ {NN}- j ]^(-1) for j = 0,..,{NN}')
for k in range(NN):
	print(k, prod([rp[i] for i in range(NN)]+ [fp[k]] +[rpi[i] for i in range(NN-1,-1,-1)]) == dpi[NN-1-k])
#############################################################
print(f'Test d[i=0,..{NN}]. f[ j ]. d[i={NN},..,0]^(-1) = r[ j ] for j=0,..,{NN}')
for k in range(NN):
	print(k, prod([dp[i] for i in range(NN)]+ [fp[k]] +[dpi[i] for i in range(NN-1,-1,-1)]) == rp[k])
#############################################################
print(f'Test d[i=0,..{NN}]. r[ j ]. d[i={NN},..,0]^(-1) = f[{NN}- j ]^(-1) for j= 0,..,{NN}')
for k in range(NN):
	print(k, prod([dp[i] for i in range(NN)]+ [rp[k]] +[dpi[i] for i in range(NN-1,-1,-1)])== fpi[NN-1-k])


 CHECKING IDENTITIES
Test f[i=0, .. 33]. r[ j ]. f[i=33,..,0]^(-1) = d[ j ] for j = 0,..,33
0 True
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True
9 True
10 True
11 True
12 True
13 True
14 True
15 True
16 True
17 True
18 True
19 True
20 True
21 True
22 True
23 True
24 True
25 True
26 True
27 True
28 True
29 True
30 True
31 True
32 True
Test f[i=0,..33]. d[ j ]. f[i=33,..,0]^(-1) = r[ 33-j ]^(-1) for j=0,..,33
0 True
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True
9 True
10 True
11 True
12 True
13 True
14 True
15 True
16 True
17 True
18 True
19 True
20 True
21 True
22 True
23 True
24 True
25 True
26 True
27 True
28 True
29 True
30 True
31 True
32 True
Test r[i=0,..33]. d[ j ]. r[i=33,..,0]^(-1) = f[ j ] for j = 0,..,33
0 True
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True
9 True
10 True
11 True
12 True
13 True
14 True
15 True
16 True
17 True
18 True
19 True
20 True
21 True
22 True
23 True
24 True
25 True
26 True
27 True
28 True
29 True
30 True
31 True
32 True
T

In [12]:
#############################################################
print('\n CHECKING IDENTITIES STARTING WITH INVERSES')
print(f'Test f[i={NN},..,0]^(-1).d[ j ]^(-1). f[i=0, .. {NN}]  = r[ j ] ^(-1) for j = 0,..,{NN}')
for k in range(NN):
    print(k, prod([fpi[i] for i in range(NN-1,-1,-1)]+ [dpi[k]] +[fp[i] for i in range(NN)]) == rpi[k])
#############################################################
print(f'Test f[i={NN},..,0]^(-1). r[ {NN}-j ]. f[i=0,..{NN}] =  d[ j ]^(-1)  for j=0,..,{NN}')
for k in range(NN):
    print(k, prod([fpi[i] for i in range(NN-1,-1,-1)]+ [rp[NN-1-k]] +[fp[i] for i in range(NN)]) == dpi[k])

#############################################################
print(f'Test  r[i={NN},..,0]^(-1). f[ j ]^(-1). r[i=0,..{NN}]=  d[ j ]^(-1)  for j = 0,..,{NN}')
for k in range(NN):
    print(k, prod([rpi[i] for i in range(NN-1,-1,-1)]+ [fpi[k]] +[rp[i] for i in range(NN)]) == dpi[k])

#############################################################
print(f'Test  r[i={NN},..,0]^(-1) d[ {NN}- j ]. r[i=0,..{NN}] = f[ j ]^(-1) for j = 0,..,{NN}')
for k in range(NN):
    print(k, prod([rpi[i] for i in range(NN-1,-1,-1)]+ [dp[NN-1-k]] +[rp[i] for i in range(NN)]) == fpi[k])
#############################################################
print(f'Test  d[i={NN},..,0]^(-1)  r[ j ]^(-1) d[i=0,..{NN}]= f[ j ]^(-1) for j=0,..,{NN}')
for k in range(NN):
    print(k, prod([dpi[i] for i in range(NN-1,-1,-1)]+ [rpi[k]] +[dp[i] for i in range(NN)]) == fpi[k])
#############################################################
print(f'Test  d[i={NN},..,0]^(-1).f[{NN}- j ]. d[i=0,..{NN}] = r[ j ]^(-1) for j= 0,..,{NN}')
for k in range(NN):
    print(k, prod([dpi[i] for i in range(NN-1,-1,-1)]+ [fp[NN-1-k]] +[dp[i] for i in range(NN)])== rpi[k])


 CHECKING IDENTITIES STARTING WITH INVERSES
Test f[i=33,..,0]^(-1).d[ j ]^(-1). f[i=0, .. 33]  = r[ j ] ^(-1) for j = 0,..,33
0 True
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True
9 True
10 True
11 True
12 True
13 True
14 True
15 True
16 True
17 True
18 True
19 True
20 True
21 True
22 True
23 True
24 True
25 True
26 True
27 True
28 True
29 True
30 True
31 True
32 True
Test f[i=33,..,0]^(-1). r[ 33-j ]. f[i=0,..33] =  d[ j ]^(-1)  for j=0,..,33
0 True
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True
9 True
10 True
11 True
12 True
13 True
14 True
15 True
16 True
17 True
18 True
19 True
20 True
21 True
22 True
23 True
24 True
25 True
26 True
27 True
28 True
29 True
30 True
31 True
32 True
Test  r[i=33,..,0]^(-1). f[ j ]^(-1). r[i=0,..33]=  d[ j ]^(-1)  for j = 0,..,33
0 True
1 True
2 True
3 True
4 True
5 True
6 True
7 True
8 True
9 True
10 True
11 True
12 True
13 True
14 True
15 True
16 True
17 True
18 True
19 True
20 True
21 True
22 True
23 True
24 True
25 True
26 True
2

# Other N

By changing the input value of $N$, the identities satisfied by $f,r,d$ operators and their inverses can be verified for other $N$ by re-evaluating the cells above.